onnx/funconnx/op_run.py

```diff
@@ -0,0 +1,508 @@
+                f"domain {self.onnx_node.domain!r}\n{att}."
+            )
+        raise AttributeError(
+            f"Unable to convert default value for {ref_att.name!r} type {att.type!r} "
+            f"from node type {self.onnx_node.op_type!r}, "
+            f"domain {self.onnx_node.domain!r}\n{att}\n{ref_att}."
+        )
+
+    def _load_attributes(self) -> None:
+        "Checks and loads attributes."
+        self.has_linked_attribute = False
+        added_attributes = []
+        for att in self.onnx_node.attribute:
+            name = att.name
+            if att.ref_attr_name:
+                value = RefAttrName(att.ref_attr_name)
+                self.has_linked_attribute = True
+            else:
+                value = self._extract_attribute_value(att)
+            setattr(self, name, value)
+            added_attributes.append(name)
+            if att.type == AttributeProto.GRAPH:
+                setattr(
+                    self,
+                    f"_run_{att.name}",
+                    lambda context, value=value: value.run(None, context or {}),
+                )
+
+        if self._schema and self.onnx_node.op_type not in {"Constant"}:
+            for k, v in self._schema.attributes.items():  # type: ignore
+                if not hasattr(self, k):
+                    if getattr(v, "required", True):
+                        raise RuntimeError(
+                            f"Attribute {k!r} is expected based on ONNX specifications "
+                            f"for node {self.onnx_node.op_type!r}."
+                        )
+                    if hasattr(v, "default_value"):
+                        if v.default_value.type == 0:
+                            # default value is undefined, it depends on the inputs
+                            value = None  # type: ignore
+                        else:
+                            value = self._extract_attribute_value(v.default_value, v)
+                        setattr(self, k, value)
+                        added_attributes.append(k)
+        self.attributes_names_ = set(added_attributes)
+
+    @staticmethod
+    def local_inputs(graph: GraphProto) -> List[str]:
+        """
+        Returns all varibles not registered as inputs and not produced by
+        an node inside the graph. This inputs are part of the context
+        existing in the graph calling this one.
+        """
+        if not isinstance(graph, GraphProto):
+            raise TypeError(f"Unexpected type {type(graph)!r}.")
+        local = set()
+        known = set()
+        for init in graph.initializer:
+            known.add(init.name)
+        for sparse_init in graph.sparse_initializer:
+            known.add(sparse_init.name)  # type: ignore
+        for inp in graph.input:
+            known.add(inp.name)
+        for node in graph.node:
+            for o in node.output:
+                known.add(o)
+            for i in node.input:
```

Comment:
Currently, this does not handle nested graph-attributes, which could also contain references to outer-outer-scope.